saturation.go

```diff
@@ -0,0 +1,148 @@
+package raft
+
+import (
+	"math"
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+// saturationMetric measures the saturation (percentage of time spent working vs
+// waiting for work) of an event processing loop, such as runFSM. It reports the
+// saturation as a gauge metric (at most) once every reportInterval.
+//
+// Callers must instrument their loop with calls to sleeping and working, starting
+// with a call to sleeping.
+//
+// Note: the caller must be single-threaded and saturationMetric is not safe for
+// concurrent use by multiple goroutines.
+type saturationMetric struct {
+	reportInterval time.Duration
+
+	// buckets contains measurements for both the current, and a configurable
+	// number of previous periods (to smooth out spikes).
+	buckets       []saturationMetricBucket
+	currentBucket int
+
+	sleepBegan, workBegan time.Time
+
+	// These are overwritten in tests.
+	nowFn    func() time.Time
+	reportFn func(float32)
+}
+
+// saturationMetricBucket contains the measurements for a period.
+type saturationMetricBucket struct {
+	// beginning of the period for which this bucket contains measurements.
+	beginning time.Time
+
+	// slept contains time for which the event processing loop was sleeping rather
+	// than working.
+	slept time.Duration
+
+	// lost contains time that is considered lost due to incorrect use
+	// of saturationMetricBucket (e.g. calling sleeping() or working() 
+	// multiple times in succession).
+	lost time.Duration
+}
+
+// newSaturationMetric creates a saturationMetric that will update the gauge
+// with the given name at the given reportInterval. keepPrev determines the
+// number of previous measurements that will be used to smooth out spikes.
+func newSaturationMetric(name []string, reportInterval time.Duration, keepPrev uint) *saturationMetric {
+	m := &saturationMetric{
+		reportInterval: reportInterval,
+		buckets:        make([]saturationMetricBucket, 1+keepPrev),
+		nowFn:          time.Now,
+		reportFn:       func(sat float32) { metrics.SetGauge(name, sat) },
```

Comment:
```suggestion
		reportFn:       func(sat float32) { metrics.AddSample(name, sat) },
```

Using `AddSample` will allow us to do aggregations over time. So if we are scraping metrics every 10s we don't just get the saturation for the last second but rather get  the min/max/mean/count sorts of metrics.

Reply:
Nice! Thanks for the recommendation, pushing this into go-metrics would remove a lot of code 🙌🏻